File: ios/Classes/ZennopayFlutterPlugin.swift
import Flutter
import UIKit
import Zennopay

/// Flutter plugin bridging Dart → the native Zennopay iOS PaymentSheet.
///
/// This is a THIN native bridge: it wraps `Zennopay.presentCheckout(...)` from
/// the native iOS SDK and resolves the Dart `present` call exactly once with a
/// map-encoded `PaymentResult`. It renders no UI itself — the native SDK owns
/// scan / amount / confirm / status (and the platform accessibility that comes
/// with it).
///
/// The host `refreshSession` hook is serviced by calling back into Dart over
/// the same channel (`refreshSession`) and awaiting the reply, so the native
/// SDK's async refresh never blocks the platform thread.
public final class ZennopayFlutterPlugin: NSObject, FlutterPlugin {

  private let channel: FlutterMethodChannel

  private init(channel: FlutterMethodChannel) {
    self.channel = channel
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "zennopay_flutter",
      binaryMessenger: registrar.messenger()
    )
    let instance = ZennopayFlutterPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "present":
      present(call: call, result: result)
    case "presentReceipt":
      presentReceipt(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - present

  private func present(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard
      let args = call.arguments as? [String: Any],
      let intentId = args["intentId"] as? String,
      let sessionJwt = args["sessionJwt"] as? String
    else {
      result(FlutterError(
        code: "invalid_jwt",
        message: "present requires intentId + sessionJwt.",
        details: nil
      ))
      return
    }

    let configMap = args["config"] as? [String: Any] ?? [:]
    let appearanceMap = args["appearance"] as? [String: Any] ?? [:]
    let config = ZennopayCodec.config(from: configMap)
    let appearance = ZennopayCodec.appearance(from: appearanceMap)

    Task { @MainActor in
      guard let presenter = ZennopayFlutterPlugin.topViewController() else {
        result(FlutterError(
          code: "network_error",
          message: "No UIViewController available to present the Zennopay sheet.",
          details: nil
        ))
        return
      }

      Zennopay.presentCheckout(
        from: presenter,
        intentID: intentId,
        sessionJWT: sessionJwt,
        refreshSession: { [weak self] intent in
          await self?.requestRefreshedSession(for: intent)
        },
        appearance: appearance,
        config: config
      ) { paymentResult in
        result(ZennopayCodec.map(from: paymentResult))
      }
    }
  }

  // MARK: - presentReceipt

  private func presentReceipt(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard
      let args = call.arguments as? [String: Any],
      let intentId = args["intentId"] as? String,
      let receiptToken = args["receiptToken"] as? String
    else {
      result(FlutterError(
        code: "invalid_jwt",
        message: "presentReceipt requires intentId + receiptToken.",
        details: nil
      ))
      return
    }

    let configMap = args["config"] as? [String: Any] ?? [:]
    let appearanceMap = args["appearance"] as? [String: Any] ?? [:]
    let config = ZennopayCodec.config(from: configMap)
    let appearance = ZennopayCodec.appearance(from: appearanceMap)

    Task { @MainActor in
      guard let presenter = ZennopayFlutterPlugin.topViewController() else {
        result(FlutterError(
          code: "network_error",
          message: "No UIViewController available to present the Zennopay receipt.",
          details: nil
        ))
        return
      }

      Zennopay.presentReceipt(
        from: presenter,
        intentID: intentId,
        receiptToken: receiptToken,
        refreshReceiptToken: { [weak self] intent in
          await self?.requestRefreshedReceiptToken(for: intent)
        },
        config: config,
        appearance: appearance
      ) {
        // The receipt is a read-only surface: it resolves the Dart call with
        // no value once the user dismisses it.
        result(nil)
      }
    }
  }

  // MARK: - refreshSession round-trip (native → Dart)

  private func requestRefreshedSession(for intentId: String) async -> String? {
    await requestRefreshedToken(method: "refreshSession", intentId: intentId)
  }

  /// Fired by the native SDK on a 401 mid-poll on the receipt: ask Dart for a
  /// fresh receipt token.
  private func requestRefreshedReceiptToken(for intentId: String) async -> String? {
    await requestRefreshedToken(method: "refreshReceiptToken", intentId: intentId)
  }

  private func requestRefreshedToken(method: String, intentId: String) async -> String? {
    await withCheckedContinuation { continuation in
      DispatchQueue.main.async {
        self.channel.invokeMethod(
          method,
          arguments: ["intentId": intentId]
        ) { reply in
          continuation.resume(returning: reply as? String)
        }
      }
    }
  }

  // MARK: - Top view controller

  @MainActor
  private static func topViewController() -> UIViewController? {
    let scenes = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
    let keyWindow = scenes
      .flatMap { $0.windows }
      .first { $0.isKeyWindow }
      ?? scenes.first?.windows.first
    var top = keyWindow?.rootViewController
    while let presented = top?.presentedViewController {
      top = presented
    }
    return top
  }
}

/// Translates the channel maps to/from the native SDK's value types.
enum ZennopayCodec {

  // MARK: config

  static func config(from map: [String: Any]) -> ZennopayConfig {
    let base = (map["apiBaseUrl"] as? String)
      .flatMap(URL.init(string:)) ?? ZennopayConfig.staging.apiBaseURL
    let pollTimeout = (map["statusPollTimeoutSeconds"] as? NSNumber)?.doubleValue ?? 90
    let maxInterval = (map["maxPollIntervalSeconds"] as? NSNumber)?.doubleValue ?? 4
    let quoteTtl = (map["defaultQuoteTtlSeconds"] as? NSNumber)?.doubleValue ?? 30
    return ZennopayConfig(
      apiBaseURL: base,
      statusPollTimeout: pollTimeout,
      maxPollInterval: maxInterval,
      defaultQuoteTTL: quoteTtl
    )
  }

  // MARK: appearance

  static func appearance(from map: [String: Any]) -> ZennopayAppearance {
    let colorsMap = map["colors"] as? [String: Any] ?? [:]
    let darkMap = colorsMap["dark"] as? [String: Any] ?? [:]

    func color(_ key: String, _ fallback: UInt32) -> UIColor {
      let light = hex(colorsMap[key] as? String) ?? fallback
      let dark = hex(darkMap[key] as? String) ?? light
      return UIColor(zpLight: light, zpDark: dark)
    }

    let colors = ZennopayAppearance.Colors(
      primary: color("primary", 0x1B6B2F),
      background: color("background", 0xFAFAF8),
      surface: color("surface", 0xFFFFFF),
      textPrimary: color("textPrimary", 0x0A0F14),
      textSecondary: color("textSecondary", 0x5A6675),
      textTertiary: color("textTertiary", 0x687280),
      border: color("border", 0xE8E9EC),
      success: color("success", 0x15803D),
      pending: color("pending", 0x7C5E1A),
      failure: color("failure", 0xA53939)
    )

    let radiusMap = map["cornerRadius"] as? [String: Any] ?? [:]
    let cornerRadius = ZennopayAppearance.CornerRadius(
      input: cg(radiusMap["input"]) ?? 4,
      card: cg(radiusMap["card"]) ?? 8,
      slide: cg(radiusMap["slide"]) ?? 12
    )

    let fontMap = map["font"] as? [String: Any] ?? [:]
    let font = ZennopayAppearance.Font(
      family: fontMap["family"] as? String ?? "General Sans",
      scale: cg(fontMap["scale"]) ?? 1
    )

    let buttonMap = map["primaryButton"] as? [String: Any] ?? [:]
    let primaryButton = ZennopayAppearance.PrimaryButton(
      background: UIColor(zpRGB: hex(buttonMap["background"] as? String) ?? 0x1B6B2F),
      textColor: UIColor(zpRGB: hex(buttonMap["textColor"] as? String) ?? 0xFFFFFF),
      cornerRadius: cg(buttonMap["cornerRadius"]) ?? 8
    )

    let mode: ZennopayAppearance.Mode
    switch map["mode"] as? String {
    case "light": mode = .light
    case "dark": mode = .dark
    default: mode = .automatic
    }

    let logo = (map["logo"] as? String).flatMap { UIImage(named: $0) }

    return ZennopayAppearance(
      mode: mode,
      colors: colors,
      cornerRadius: cornerRadius,
      font: font,
      primaryButton: primaryButton,
      logo: logo
    )
  }

  // MARK: result

  static func map(from result: PaymentResult) -> [String: Any] {
    switch result {
    case let .completed(intentID):
      return ["status": "completed", "intentId": intentID]
    case let .pending(intentID):
      return ["status": "pending", "intentId": intentID]
    case let .canceled(intentID):
      return ["status": "canceled", "intentId": intentID]
    case let .failed(intentID, error):
      return [
        "status": "failed",
        "intentId": intentID,
        "error": ["code": wireCode(error)],
      ]
    }
  }

  /// Collapse the native iOS `ZennopayError` onto the stable Flutter wire codes
  /// (`ZennopayErrorCode.wire`) so Dart's `fromWire` resolves them directly.
  private static func wireCode(_ error: ZennopayError) -> String {
    switch error {
    case .invalidJWT, .malformedToken, .jwtMissingClaim, .presentationContextMissing:
      return "invalid_jwt"
    case .intentMismatch:
      return "intent_mismatch"
    case .jwtExpired, .sessionExpired:
      return "jwt_expired"
    case .confirmReplay:
      return "jti_replay"
    case .invalidQRCode:
      return "qr_invalid"
    case .quoteExpired:
      return "quote_expired"
    case .paymentFailed:
      return "confirm_failed"
    case .userCanceled:
      return "canceled"
    case .cameraPermissionDenied:
      return "camera_denied"
    case .timedOut:
      return "timed_out"
    case .networkError, .serverError:
      return "network_error"
    }
  }

  // MARK: helpers

  private static func hex(_ string: String?) -> UInt32? {
    guard var s = string else { return nil }
    if s.hasPrefix("#") { s.removeFirst() }
    return UInt32(s, radix: 16)
  }

  private static func cg(_ value: Any?) -> CGFloat? {
    (value as? NSNumber).map { CGFloat(truncating: $0) }
  }
}

File: android/src/main/kotlin/in/zennopay/flutter/ZennopayFlutterPlugin.kt
package `in`.zennopay.flutter

import android.os.Handler
import android.os.Looper
import androidx.activity.ComponentActivity
import androidx.compose.ui.unit.dp
import com.zennopay.sdk.PaymentResult
import com.zennopay.sdk.Zennopay
import com.zennopay.sdk.ZennopayConfig
import com.zennopay.sdk.ui.ZennopayAppearance
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume

/**
 * Flutter plugin bridging Dart → the native Zennopay Android PaymentSheet.
 *
 * A THIN native bridge: it wraps `com.zennopay.sdk.Zennopay.presentCheckout(...)`
 * against the current Activity and resolves the Dart `present` call exactly once
 * with a map-encoded [PaymentResult]. It renders no UI itself — the native SDK
 * owns scan / amount / confirm / status (and the platform accessibility that
 * comes with it).
 *
 * The host `refreshSession` hook is serviced by calling back into Dart over the
 * same channel (`refreshSession`) and awaiting the reply, so the native SDK's
 * `suspend` refresh never blocks.
 */
class ZennopayFlutterPlugin : FlutterPlugin, ActivityAware, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private val mainHandler = Handler(Looper.getMainLooper())

    private var activity: ComponentActivity? = null

    /** The Flutter result for the in-flight `present` call. Single-shot. */
    private var pendingResult: Result? = null

    /** The Flutter result for the in-flight `presentReceipt` call. Single-shot. */
    private var pendingReceiptResult: Result? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "zennopay_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // ---- ActivityAware --------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as? ComponentActivity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity as? ComponentActivity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // ---- MethodCallHandler ----------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "present" -> present(call, result)
            "presentReceipt" -> presentReceipt(call, result)
            else -> result.notImplemented()
        }
    }

    private fun present(call: MethodCall, result: Result) {
        val intentId = call.argument<String>("intentId")
        val sessionJwt = call.argument<String>("sessionJwt")
        if (intentId == null || sessionJwt == null) {
            result.error("invalid_jwt", "present requires intentId + sessionJwt.", null)
            return
        }

        val host = activity
        if (host == null) {
            result.error(
                "network_error",
                "No ComponentActivity available to present the Zennopay sheet.",
                null,
            )
            return
        }

        val config = ZennopayCodec.config(call.argument<Map<String, Any?>>("config"))
        val appearance = ZennopayCodec.appearance(call.argument<Map<String, Any?>>("appearance"))

        pendingResult = result

        Zennopay.presentCheckout(
            activity = host,
            intentId = intentId,
            sessionJwt = sessionJwt,
            refreshSession = { intent -> requestRefreshedSession(intent) },
            appearance = appearance,
            config = config,
        ) { paymentResult ->
            mainHandler.post {
                pendingResult?.success(ZennopayCodec.map(paymentResult))
                pendingResult = null
            }
        }
    }

    private fun presentReceipt(call: MethodCall, result: Result) {
        val intentId = call.argument<String>("intentId")
        val receiptToken = call.argument<String>("receiptToken")
        if (intentId == null || receiptToken == null) {
            result.error("invalid_jwt", "presentReceipt requires intentId + receiptToken.", null)
            return
        }

        val host = activity
        if (host == null) {
            result.error(
                "network_error",
                "No ComponentActivity available to present the Zennopay receipt.",
                null,
            )
            return
        }

        val config = ZennopayCodec.config(call.argument<Map<String, Any?>>("config"))
        val appearance = ZennopayCodec.appearance(call.argument<Map<String, Any?>>("appearance"))

        pendingReceiptResult = result

        Zennopay.presentReceipt(
            activity = host,
            intentId = intentId,
            receiptToken = receiptToken,
            refreshReceiptToken = { intent -> requestRefreshedReceiptToken(intent) },
            config = config,
            appearance = appearance,
        ) {
            // The receipt is a read-only surface: resolve the Dart call with no
            // value once the user dismisses it.
            mainHandler.post {
                pendingReceiptResult?.success(null)
                pendingReceiptResult = null
            }
        }
    }

    /** Fired by the native SDK on 401/expiry: ask Dart for a fresh JWT. */
    private suspend fun requestRefreshedSession(intentId: String): String? =
        requestRefreshedToken("refreshSession", intentId)

    /** Fired by the native SDK on a 401 mid-poll on the receipt: ask Dart for a
     * fresh receipt token. */
    private suspend fun requestRefreshedReceiptToken(intentId: String): String? =
        requestRefreshedToken("refreshReceiptToken", intentId)

    private suspend fun requestRefreshedToken(method: String, intentId: String): String? =
        suspendCancellableCoroutine { continuation ->
            mainHandler.post {
                channel.invokeMethod(
                    method,
                    mapOf("intentId" to intentId),
                    object : Result {
                        override fun success(result: Any?) =
                            continuation.resume(result as? String)

                        override fun error(code: String, message: String?, details: Any?) =
                            continuation.resume(null)

                        override fun notImplemented() =
                            continuation.resume(null)
                    },
                )
            }
        }
}

/** Translates the channel maps to/from the native SDK's value types. */
private object ZennopayCodec {

    fun config(map: Map<String, Any?>?): ZennopayConfig {
        val m = map ?: emptyMap()
        val environment = when (m["environment"] as? String) {
            "production" -> ZennopayConfig.Environment.PRODUCTION
            "custom" -> ZennopayConfig.Environment.CUSTOM
            else -> ZennopayConfig.Environment.STAGING
        }
        val base = (m["apiBaseUrl"] as? String)
            ?: ZennopayConfig.DEFAULT_STAGING_BASE_URL
        val timeoutMillis =
            ((m["maxPollIntervalSeconds"] as? Number)?.toLong()?.times(1000)) ?: 20_000L
        return ZennopayConfig(
            apiBaseUrl = base,
            environment = environment,
            requestTimeoutMillis = timeoutMillis.coerceAtLeast(1_000L),
        )
    }

    fun appearance(map: Map<String, Any?>?): ZennopayAppearance {
        val m = map ?: return ZennopayAppearance.Automatic
        @Suppress("UNCHECKED_CAST")
        val colorsMap = m["colors"] as? Map<String, Any?> ?: emptyMap()

        val colors = ZennopayAppearance.Colors(
            primary = argb(colorsMap["primary"], 0xFF1B6B2F),
            background = argb(colorsMap["background"], 0xFFFAFAF8),
            surface = argb(colorsMap["surface"], 0xFFFFFFFF),
            textPrimary = argb(colorsMap["textPrimary"], 0xFF0A0F14),
            textSecondary = argb(colorsMap["textSecondary"], 0xFF5A6675),
            textTertiary = argb(colorsMap["textTertiary"], 0xFF687280),
            border = argb(colorsMap["border"], 0xFFE8E9EC),
            success = argb(colorsMap["success"], 0xFF15803D),
            pending = argb(colorsMap["pending"], 0xFF7C5E1A),
            failure = argb(colorsMap["failure"], 0xFFA53939),
        )

        @Suppress("UNCHECKED_CAST")
        val radius = m["cornerRadius"] as? Map<String, Any?> ?: emptyMap()
        val shapes = ZennopayAppearance.Shapes(
            input = ((radius["input"] as? Number)?.toDouble() ?: 4.0).dp,
            card = ((radius["card"] as? Number)?.toDouble() ?: 8.0).dp,
            slide = ((radius["slide"] as? Number)?.toDouble() ?: 12.0).dp,
        )

        @Suppress("UNCHECKED_CAST")
        val font = m["font"] as? Map<String, Any?> ?: emptyMap()
        val typography = ZennopayAppearance.Typography(
            // Arbitrary partner font families can't be resolved to a Compose
            // FontFamily here; the native SDK falls back to its bundled sans.
            fontFamily = null,
            scale = (font["scale"] as? Number)?.toFloat() ?: 1f,
        )

        @Suppress("UNCHECKED_CAST")
        val button = m["primaryButton"] as? Map<String, Any?> ?: emptyMap()
        val primaryButton = ZennopayAppearance.PrimaryButton(
            background = argb(button["background"], 0xFF1B6B2F),
            textColor = argb(button["textColor"], 0xFFFFFFFF),
            cornerRadius = ((button["cornerRadius"] as? Number)?.toDouble() ?: 8.0).dp,
        )

        val mode = when (m["mode"] as? String) {
            "light" -> ZennopayAppearance.Mode.Light
            "dark" -> ZennopayAppearance.Mode.Dark
            else -> ZennopayAppearance.Mode.Automatic
        }

        return ZennopayAppearance(
            mode = mode,
            colors = colors,
            shapes = shapes,
            typography = typography,
            primaryButton = primaryButton,
            logo = null,
        )
    }

    fun map(result: PaymentResult): Map<String, Any?> = when (result) {
        is PaymentResult.Completed -> mapOf(
            "status" to "completed",
            "intentId" to result.intentId,
            "receipt" to mapOfNonNull(
                "merchantName" to result.merchantName,
                "localAmount" to result.localAmount,
                "localCurrency" to result.localCurrency,
                "usdDebited" to result.usdDebited,
                "transactionId" to result.transactionId,
                "verifiableQrData" to result.verifiableQrData,
            ),
        )
        is PaymentResult.Pending -> mapOf(
            "status" to "pending",
            "intentId" to result.intentId,
        )
        is PaymentResult.Canceled -> mapOf(
            "status" to "canceled",
            "intentId" to (result.intentId ?: ""),
        )
        is PaymentResult.Failed -> mapOf(
            "status" to "failed",
            "intentId" to (result.intentId ?: ""),
            "error" to mapOf("code" to wireCode(result.error.code)),
        )
    }

    /**
     * Collapse the native Android dotted error codes onto the stable Flutter
     * wire codes (`ZennopayErrorCode.wire`) so Dart's `fromWire` resolves them.
     */
    private fun wireCode(code: String): String = when (code) {
        "client.invalid_jwt", "client.malformed_token",
        "client.invalid_issuer", "client.missing_intent_id" -> "invalid_jwt"
        "client.intent_mismatch", "jwt.intent_id_mismatch_with_path" -> "intent_mismatch"
        "client.jwt_expired", "auth.refresh_failed", "auth.unauthorized" -> "jwt_expired"
        "jwt.jti_replay" -> "jti_replay"
        "scanner.camera_denied" -> "camera_denied"
        "scanner.qr_undecodable", "scan.validation_failed" -> "qr_invalid"
        "confirm.quote_expired", "confirm.quote_mismatch",
        "confirm.quote_superseded" -> "quote_expired"
        "confirm.dynamic_amount_override" -> "amount_not_allowed"
        "confirm.not_scanned", "jwt.intent_invalid_state",
        "payment.declined" -> "confirm_failed"
        "status.polling_timeout" -> "timed_out"
        "network.error" -> "network_error"
        else -> "network_error"
    }

    private fun mapOfNonNull(vararg pairs: Pair<String, Any?>): Map<String, Any?>? {
        val filtered = pairs.filter { it.second != null }
        return if (filtered.isEmpty()) null else filtered.toMap()
    }

    private fun argb(value: Any?, fallback: Long): Long {
        val hex = value as? String ?: return fallback
        val rgb = hex.removePrefix("#").toLongOrNull(16) ?: return fallback
        return 0xFF000000L or (rgb and 0xFFFFFF)
    }
}
